lib/sentry/client_report.ex

```diff
@@ -0,0 +1,128 @@
+defmodule Sentry.ClientReport do
+  @moduledoc """
+  **TODO** add proper doc for module
+
+  See <https://develop.sentry.dev/sdk/client-reports/>.
+  """
+
+  @moduledoc since: "10.0.0"
+
+  @typedoc since: "10.0.0"
+
+  use GenServer
+  alias Sentry.Client
+
+  @typedoc """
+  The possible reasons of the discarded event.
+  """
+  @type reasons() ::
+          :ratelimit_backoff
+          | :queue_overflow
+          | :cache_overflow
+          | :network_error
+          | :sample_rate
+          | :before_send
+          | :event_processor
+          | :insufficient_data
+          | :backpressure
+
+  @typedoc """
+  The struct for a **client report** interface.
+  """
+  @type t() :: %__MODULE__{
+          timestamp: String.t() | number(),
+          discarded_events:
+            list(%{reason: reasons(), category: String.t(), quantity: pos_integer()})
+        }
+
+  defstruct [:timestamp, :discarded_events]
+
+  @client_report_reasons [
+    :ratelimit_backoff,
+    :queue_overflow,
+    :cache_overflow,
+    :network_error,
+    :sample_rate,
+    :before_send,
+    :event_processor,
+    :insufficient_data,
+    :backpressure
+  ]
+
+  @send_interval 5000
+
+  @spec start_link(keyword()) :: GenServer.on_start()
+  def start_link(_) do
+    # check config to see if send_client_report is true
+    GenServer.start_link(__MODULE__, %__MODULE__{}, name: __MODULE__)
+  end
+
+  @spec add_discarded_event({reasons(), String.t()}) :: :ok
+  def add_discarded_event({reason, category}) do
+    if Enum.member?(@client_report_reasons, reason) do
+      GenServer.cast(__MODULE__, {:add_discarded_event, {reason, category}})
+    end
+
+    :ok
+  end
```

Comment:
Why did you choose to silently ignore events whose `reason` is not a valid one?

Reply:
I decided to do this for two reasons -> From what I can tell, the Ruby SDK ignores invalid reasons 

``` ruby
    def record_lost_event(reason, data_category, num: 1)
      return unless @send_client_reports
      return unless CLIENT_REPORT_REASONS.include?(reason)

      @discarded_events[[reason, data_category]] += num
    end   
```

and the documentation says 

`In case a reason needs to be added, it also has to be added to the allowlist in [snuba](https://github.com/getsentry/snuba/blob/4e7cfdddcf7b93eacb762bc74ca2461cec9464e5/snuba/datasets/outcomes_processor.py#L24-L34).`
    
So we choose to do nothing if it's invalid unless we want to add them to this repo mentioned. Thoughts?

Reply:
Okay, sounds good. Let's silently ignore it for now, BUT leave a comment with basically what you just said to me, but in the code 🙃